eval/adjust.rkt

```diff
@@ -71,333 +71,263 @@
     (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
     (define output (vector-ref vregs n))                            ; output of the current instr
     
-    (define exps-from-above (vector-ref vprecs-new (- n varc)))     ; vprecs-new is shifted by varc elements from vregs
-    (define new-exponents (get-exponent op output srcs))
+    (define intro (vector-ref vprecs-new (- n varc)))               ; intro for the current instruction
+    (define ampls (get-ampls op output srcs))                       ; ampls for the tail instructions
 
-    (define final-parent-precision (max (+ exps-from-above
+    (define final-parent-precision (max (+ intro
                                            (vector-ref vstart-precs (- n varc)))
                                         (*base-tuning-precision*)))
-
-    ; This case is weird. if we have a cancellation in fma -> ival-mult in fma should be in higher precision
-    (when (equal? op ival-fma)
-      (set! final-parent-precision (+ final-parent-precision (car new-exponents))))
     
-    (when (>= final-parent-precision (*rival-max-precision*))         ; Early stopping
+    (when (>= final-parent-precision (*rival-max-precision*))       ; Early stopping
       (*sampling-iteration* (*rival-max-iterations*)))
+    
+    ; Final precision assignment
     (vector-set! vprecs-new (- n varc) (min final-parent-precision (*rival-max-precision*)))
 
+    ; Intro and ampl propogation for each tail instruction
     (for ([x (in-list tail-registers)]
-          [new-exp (in-list new-exponents)]
+          [ampl (in-list ampls)]
           #:when (>= x varc)) ; when tail register is not a variable
       ; check whether this op already has a precision that is higher
-      (when (> (+ exps-from-above new-exp) (vector-ref vprecs-new (- x varc)))
-        (vector-set! vprecs-new (- x varc) (+ exps-from-above new-exp))))))
+      (when (> (+ intro ampl) (vector-ref vprecs-new (- x varc)))
+        (vector-set! vprecs-new (- x varc) (+ intro ampl))))))
+
+(define (exp-is-inf? x)
+  (equal? x 9223372036854775805))
+(define (exp-is-0? x)
+  (equal? x -9223372036854775807))
+; this function imitates an exponent of infinity as 9223372036854775805
+(define (replace-inf x)
+  (if (equal? x -9223372036854775805)
+      (abs x)
+      x))
+(define (crosses-zero? x)
+  (not (equal? (mpfr-sign (ival-lo x)) (mpfr-sign (ival-hi x)))))
```

Comment:
I'm not sure if this is quite right because there's also the case that `lo = hi = 0.0` or `lo = hi = -0.0`. I think these come up rarely and don't affect our results (because these are point intervals which probably mean that things are good) but we might try to add them to the logic.